### PR TITLE
Compare image in the container registry and one in git

### PR DESCRIPTION
**Pull request by nakabonne**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes https://github.com/pipe-cd/pipe/issues/1209

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/app/piped/imagewatcher/watcher.go [hunk 1]
```diff
@@ -96,90 +100,99 @@ func (w *watcher) run(ctx context.Context, provider imageprovider.Provider, inte
...
-				zap.Error(err),
-			)
+// determineUpdates gives back target images to be updated for a given repo.
+func (w *watcher) determineUpdates(ctx context.Context, repoID string, repo git.Repo, provider imageprovider.Provider) ([]config.ImageWatcherTarget, error) {
+	branch := repo.GetClonedBranch()
+	w.mu.Lock()
+	err := repo.Pull(ctx, branch)
+	w.mu.Unlock()
+	if err != nil {
+		return nil, fmt.Errorf("failed to update repository branch: %w", err)
```
Comment (khanhtc1202) [inline pkg/app/piped/imagewatcher/watcher.go:134]: nit: "failed to get update from branch" or "failed to pull branch"

Comment (nakabonne) [reply pkg/app/piped/imagewatcher/watcher.go:134]: Yup, you're right, thank you!

File: pkg/app/piped/imagewatcher/watcher.go [hunk 2]
```diff
@@ -96,90 +100,99 @@ func (w *watcher) run(ctx context.Context, provider imageprovider.Provider, inte
...
 
-// filterTargets gives back the targets corresponding to the given provider.
-func filterTargets(provider string, targets []config.ImageWatcherTarget) (filtered []config.ImageWatcherTarget) {
-	for _, t := range targets {
-		if t.Provider == provider {
-			filtered = append(filtered, t)
-		}
+// checkOutdated checks if the image defined in the given target is identical to the one in image provider.
+func checkOutdated(ctx context.Context, target config.ImageWatcherTarget, repo git.Repo, provider imageprovider.Provider) (bool, error) {
+	if provider.Name() != target.Provider {
```
Comment (khanhtc1202) [inline pkg/app/piped/imagewatcher/watcher.go:170]: nit: could you add notes to clarify why do we treat this case as a not outdated image 🙏

Comment (nakabonne) [reply pkg/app/piped/imagewatcher/watcher.go:170]: As you feel, it's unclear. gonna update them.

Comment (nakabonne) [reply pkg/app/piped/imagewatcher/watcher.go]: Updated them.
The Run() spawns runners for each image provider, so the checkOutdated function is called in the context for image provider. We should ignore the targets for other providers.

**Comment by khanhtc1202** [comment]
Nice 👍 
/approve

File: pkg/app/piped/imagewatcher/watcher.go (hunk 2 above)
Comment (khanhtc1202) [reply pkg/app/piped/imagewatcher/watcher.go]: 🙆

**Pull request merged**